### DISCUSSION TITLE
Way to name tracks?

### DISCUSSION THREAD

**Discussion by ffolkes** [4gra/pclk-mn10#5] [Ideas]
Thanks for your great work on this, I was able to get it running with my MXD-D40, PCLK-MN10, an Mac. Do you know how to input a track title? Is that possible? The XP client allows it, but I don't know what command it is using to do it. Thanks for any help!

**Comment by 4gra** [comment]
Thanks so much for reporting back, great to add another model to the fold. Yes, it's entirely possible to title tracks using this mechanism, but I've never gotten around to adding it, and titles are probably one of the trickier things to compose.

You can *get* titles using: `(09 00 60) B0 48 00 01 <track no> 00 <chunk>"` (where chunks and track numbers start at `01`), and you will be able to set them using something like `B0 98 00 01 <track number> 00  <chunk> <data>` but the devil's in the detail of `<data>` of course.  

Formatting ASCII into the title should be easy enough, but there are other little tricks in the encoding, and if titles exceed a single chunk there's semantics I've never looked into.

I'll dig out better records if that's not enough of a start for you -- I'd appreciate any research or achievements you make.

**Comment by ffolkes** [comment]
Thank you for the reply! I was able to get a USB sniffer running on the XP VM I use for M-Crew, and I titled a disc "testing" and stopped capturing. I have no idea how to figure out how you knew what hex does what, so I just searched for "98 00 01" and this came up. Note: this was a disc title operation, not a track title. The string "testing" was referenced several times, but only once from the PC to the deck.

Here is that packet:

0000   1b 00 48 ee 46 82 ff ff ff ff 00 00 00 00 09 00
0010   00 02 00 04 00 01 03 1a 00 00 00 19 00 60 b0 98
0020   00 01 00 00 01 74 65 73 74 69 6e 67 00 00 00 00
0030   00 00 00 00 00

"testing" is "74 65 73 74 69 6e 67" in ASCII, which starts in the middle of line 0020.

Then another packet comes from the PC after that one:

0000   1b 00 48 ee 46 82 ff ff ff ff 00 00 00 00 09 00
0010   01 02 00 04 00 01 03 00 00 00 00

I don't have enough knowledge of Python (actually, any, at all) to figure out what your script is doing, but I tried adding to your JSON a test command:

`"MDdisctitletesting":"b0 98 00 01 00 00 01 74 65 73 74 69 6e 67 00 00 00 00 00 00 00 00 00",`

But it doesn't work. My goal with all of this is not to name discs (I just thought that would be the easiest starting place), but to name tracks. I made a script that monitors VLC, and when the track changes, it sends the "MDrec" command to the deck to mark the new track. I'm hoping to eventually be able to also send a "MDnametrack" command with the current song title to make my life complete.

Any input is very much appreciated!

Edit: Some info I found from https://www.minidisc.org/faq_sec_5.html#_q45: "The disc title is handled just like any other track title (the disc title is placed on track #0, which is actually the MD's free list)."

**Comment by ffolkes** [comment]
It seems to be working now! Turns out sometimes the deck gets stuck. I'm not sure what causes it yet, perhaps the switching between VMware/XP and OS X, or maybe something M-Crew is doing. I can tell, because "off" doesn't do anything. Once I manually turn it off, "on" and "off" work again. Then when I tried "b0 98 00 01 00 00 01 74 65 73 74 69 6e 67 00 00 00 00 00 00 00 00 00" the name changed! I will try more, and see how it works.

OK, I ran another capture, this time for a disc title of "twentycharactershere" and this is what M-Crew sent:

"twentycharactershere" is ASCII "74 77 65 6e 74 79 63 68 61 72 61 63 74 65 72 73 68 65 72 65"

0000   1b 00 d8 3c ff 81 ff ff ff ff 00 00 00 00 09 00
0010   00 02 00 04 00 01 03 1a 00 00 00 19 00 60 b0 98
0020   00 01 00 00 01 74 77 65 6e "74 79 63 68 61 72 61
0030   63 74 65 72 73" (ASCII: "twentycharacters")

Followed by:

0000   1b 00 d8 3c ff 81 ff ff ff ff 00 00 00 00 09 00
0010   01 02 00 04 00 01 03 00 00 00 00

Then:

0000   1b 00 88 05 36 82 ff ff ff ff 00 00 00 00 09 00
0010   00 02 00 04 00 01 03 1a 00 00 00 19 00 60 b0 98
0020   00 01 00 00 02 68 65 72 65 00 00 00 00 00 00 00
0030   00 00 00 00 00 (ASCII: "here")

Followed by:

0000   1b 00 88 05 36 82 ff ff ff ff 00 00 00 00 09 00
0010   01 02 00 04 00 01 03 00 00 00 00

Later today I will try incrementing the chunk and sending a second packet, but if you see this before then, why is the first line of the packet different? How did you know all commands start with b0? Also, I'm not sure what those follow-up packets do, but they don't seem necessary to execute commands. Maybe M-Crew is issuing a command for itself (refresh perhaps) or something.

**Reply by 4gra** [reply]
M-Crew definitely sends (and reads) more than is required. Also note there are a display updates taking place according to some protocol (which might differ a little depending on model) - one mini-project I keep meaning to start is to keep track of the latest display state as it is updated.  M-Crew if I remember correctly obsessively checks things like volume levels to keep in sync.

> Turns out sometimes the deck gets stuck. I'm not sure what causes it yet, perhaps the switching between 
> VMware/XP and OS X, or maybe something M-Crew is doing. I can tell, because "off" doesn't do anything

Keep me posted.  If your device doesn't have the ability to power off remotely, it might just be that you're powering off the PCLK-MN10; it's easy to get that into a state where it doesn't wake up until a re-plug. But if you discover anything unique to the MXD-D40 we can record it here.

>  if you see this before then, why is the first line of the packet different? 
> How did you know all commands start with b0? 

As I say -- I don't yet understand the multi-part commands, and I'll need to program in support for this somehow. 

As to "why B0?" - that's the global "address" Sony give to MiniDisc devices / modules in the command space - well actually 'B' is the minidisc value; '9' is the CD module, etc. Messages to the MD module use `B0;` replies use `B8`. Old Control-A1 references tell me that `90-97` refer to concurrent control of up to eight CD players, so I assume `B0-B7` work the same way (not that I've ever seen a multi-MD system).  Replies use the higher 8 bits, obviously. Take a look at https://github.com/4gra/pclk-mn10/blob/master/commands.json where you should be able to see some patterns. I discuss this in the wiki but never really formalised it.  Long term project also involves parsing replies - again, just listing them would help a lot.

PS: If you haven't read it already, https://github.com/4gra/pclk-mn10/wiki/messages and the other wiki pages have a lot of basic info, though they definitely all need cleaned up as I've probably answered lots of my own questions by now.

**Comment by 4gra** [comment]
By the way these are all great additions and I'd be happy to integrate the new titling commands. I probably don't have the ability to add commands with ASCII parameters right now, so I'll add that functionality to the python.  Shouldn't take too long but in the meanwhile your research continues to be appreciated - hopefully we'll get something useful together.

**Comment by 4gra** [comment]
I've made a [bug fix](https://github.com/4gra/pclk-mn10/commit/691b57f223ca074aa3d32729f1401171ad8357a9) (so that simple numerical parameters are actually interpreted properly) and added all I presently know about [chunked messages](https://github.com/4gra/pclk-mn10/wiki/chunks) (which is to say, very little) into a wiki page.  None of it's rocket science, it just needs some light shone one it.

**Comment by ffolkes** [comment]
Thanks for all the replies! I was able to test my theory, and it works. I made two test commands, one that says "twentycharacters" and another command (with chunk incremented to 02) that says "here" and ran them sequentially, and it worked. I would imagine the deck allows about 1 second for another chunk to come before it finalizes what it has. 

I'm going to write a PHP script to format the packets for myself and play around, but if I had the ability to program python, I think you just need to pad all name operations with potentially sixteen 00's, depending on how many characters are actually used. You know what I mean - eg. "test" would be followed by twelve 00's. Converse to that, that is to say something longer than 16, needs to be broken into a second packet with an incremented chunk, and then the same padding to sixteen.

Thank you again for all your work on this! The 4 people on planet Earth who use this PCLK-MN10/20 device thank you... :P

**Comment by ffolkes** [comment]
This works for me:

`php-cgi send_title.php 0 "my disc title can be very long"`
`php-cgi send_title.php 2 "track 2 title can be very long, too!"`

Contents of send_title.php:
```php
<?php
// php-cgi thisfile.php <track number> "<title string>"

if (!isset($argv[1])) {
    exit("Track number required.");
}

if (!isset($argv[2])) {
    exit("Title string required.");
}

$arg_track = $argv[1];
$arg_title = $argv[2];

// turn every character of title into an element in an array to make things a little easier, then convert each one to hex.
$ary_title = str_split($arg_title);

// function to build the prefix based on track number and chunk increment. pad to be double-digit.
function composePrefix($track, $chunk = 1) {
    return "b0 98 00 01 " . str_pad($track, 2, "0", STR_PAD_LEFT) . " 00 " . str_pad($chunk, 2, "0", STR_PAD_LEFT) . " ";
}

// function to pad the packet with 00's if data consists of less than 16 bytes.
function padPacketWithZeros($ary_title, $offset = 0) {
    $result = "";

    // for each byte, check to see if is defined in the title array, and if not, replace with padding 00.
    // this could be done by checking character indexes instead of an array, but an array seemed faster.
    for ($i = $offset; $i < 16 + $offset; $i++){
        if (isset($ary_title[$i])) {
            // remember to anticipate an empty string if the user wants to set a blank title (aka convert "" to 00).
            if ($ary_title[$i] == "") {
                $result .= "00 ";
            } else {
                $result .= bin2hex($ary_title[$i]) . " ";
            }
        } else {
            $result .= "00 ";
        }
    }
    return trim($result);
}

// iterate through blocks of 16 bytes and send to deck. there always needs to be at least 1 iteration.
for ($i = 0; $i < ceil(count($ary_title) / 16); $i++) {
    $output = padPacketWithZeros($ary_title, 16 * $i);  // generate the data portion.
    $send_to_deck = composePrefix($arg_track, $i + 1) . "$output";  // compose the command/chunk/data portions.
    echo "sending to deck: $send_to_deck\n";
    exec("/Users/Matrix/Projects/pclk-mn10-master/stub.py asend \"$send_to_deck\"");  // write packet to deck.
}

echo "\n";
?>
```

**Reply by 4gra** [reply]
Thanks for the code - I'm amused to see PHP as a scripting language - it's super useful.

Last night I started work on the equivalent of this, to send multi-part commands with automatic padding and incrementing the chunk. 

It's not plugged in yet -- hopefully tonight, so if everything works you won't need to exec() for much longer, and will just be able to perform `MDSetTrkTitle NN [string]`.

I'll keep you posted.

**Reply by ffolkes** [reply]
I look forward to it! Is there a way to get more character support? I tried a few things in PHP, but nothing worked. When I tried to send some basic ASCII symbols, the deck displayed Japanese stuff instead. Which, ironically, is actually what I want. I lot of the vaporwave stuff I'm putting to MiniDisc has Japanese writing in the title, and since MiniDisc originated in Japan, I'd imagine the equipment must be able to handle it somehow.

I don't know why, but I've always found PHP to be very useful and easy to code in, especially things like parsing data, etc. I've been using PHP on the command-line for decades. When I used to be more involved in website hosting, I made gigantic scripts that handled various server tasks, and everyone was baffled why I didn't use bash haha.

**Reply by ffolkes** [reply]
Here is some more about that: https://www.reddit.com/r/minidisc/comments/i0tbte/katakana_titles_using_netmd_and_sonicstage/

**Reply by 4gra** [reply]
As far as I know, they support ASCII, or *Katakana (and very little else), and definitely not via Unicode. I am sure I have a reference for Japanese character entry somewhere, I'll see if I can find it.
